File: mesh/gmsh_interface.cc
#include "mesh/gmsh_interface.h"

#include <gmsh.h>

#include <algorithm>
#include <array>
#include <cstdint>
#include <iterator>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

namespace gmsh {

GmshInterface::GmshInterface() { gmsh::initialize(); }

GmshInterface::~GmshInterface() { gmsh::finalize(); }

void GmshInterface::WriteMeshFile(const std::string& mesh_file) {
  gmsh::write(mesh_file);
}

void GmshInterface::LaunchGui() { gmsh::fltk::run(); }

std::vector<Tag> GmshInterface::GetNodes(const int tag, const int dim,
                                         const NodeType node_type) {
  // Get all node tags.
  std::vector<Tag> node_tags;
  std::vector<double> node_coordinates;
  std::vector<double> node_parametric_coordinates;
  gmsh::model::mesh::getNodes(
      node_tags, node_coordinates, node_parametric_coordinates, dim, tag,
      /*includeBoundary=*/true, /*returnParametricCoord=*/false);

  // Get internal node tags.
  std::vector<Tag> internal_node_tags;
  std::vector<double> internal_node_coordinates;
  std::vector<double> internal_node_parametric_coordinates;
  gmsh::model::mesh::getNodes(internal_node_tags, internal_node_coordinates,
                              internal_node_parametric_coordinates, dim, tag,
                              /*includeBoundary=*/false,
                              /*returnParametricCoord=*/false);

  switch (node_type) {
    case NodeType::kAll: {
      return node_tags;
    }
    case NodeType::kInternal: {
      return internal_node_tags;
    }
    case NodeType::kBoundary: {
      // Get boundary node tags.
      std::unordered_set<Tag> internal_node_tags_set(
          internal_node_tags.cbegin(), internal_node_tags.cend());
      std::vector<Tag> boundary_node_tags;
      std::copy_if(node_tags.cbegin(), node_tags.cend(),
                   std::back_inserter(boundary_node_tags), [&](const Tag tag) {
                     return !internal_node_tags_set.contains(tag);
                   });
      return boundary_node_tags;
    }
    default: {
      throw std::invalid_argument("Invalid node type.");
    }
  }
}

std::unordered_map<Tag, Coordinates> GmshInterface::GetNodeCoordinates(
    const int tag, const int dim) {
  // Get all node tags and coordinates.
  std::vector<Tag> node_tags;
  std::vector<double> node_coordinates;
  std::vector<double> node_parametric_coordinates;
  gmsh::model::mesh::getNodes(
      node_tags, node_coordinates, node_parametric_coordinates, dim, tag,
      /*includeBoundary=*/true, /*returnParametricCoord=*/false);

  // Insert the node tags and coordinates into a map.
  std::unordered_map<Tag, Coordinates> node_tag_to_coordinates;
  for (uint32_t i = 0; i < node_tags.size(); ++i) {
    node_tag_to_coordinates.try_emplace(
        node_tags[i], std::array<double, 3>{node_coordinates[3 * i],
                                            node_coordinates[3 * i + 1],
                                            node_coordinates[3 * i + 2]});
  }
  return node_tag_to_coordinates;
}

std::vector<DimTag> GetBoundary(const int dim, const int tag) {
  std::vector<DimTag> input_dim_tags{{dim, tag}};
  std::vector<DimTag> output_dim_tags;
  gmsh::model::getBoundary(input_dim_tags, output_dim_tags, /*combined=*/true,
                           /*oriented=*/false, /*recursive=*/false);
  return output_dim_tags;
}

std::pair<std::vector<int>, std::vector<int>> GmshInterface::GetAdjacencies(
    const int dim, const int tag) {
  std::vector<int> upward_adjacencies;
  std::vector<int> downward_adjacencies;
  gmsh::model::getAdjacencies(dim, tag, upward_adjacencies,
                              downward_adjacencies);
  return std::make_pair(std::move(upward_adjacencies),
                        std::move(downward_adjacencies));
}

std::unordered_map<Tag, std::array<Tag, 2>> GmshInterface::GetLines(
    const int tag) {
  std::vector<Tag> line_tags;
  std::vector<Tag> node_tags;
  gmsh::model::mesh::getElementsByType(
      static_cast<int>(ElementType::kTwoNodeLine), line_tags, node_tags, tag);

  // Insert the line and node tags into a map.
  std::unordered_map<Tag, std::array<Tag, 2>> lines;
  for (uint32_t i = 0; i < line_tags.size(); ++i) {
    lines.try_emplace(
        line_tags[i],
        std::array<gmsh::Tag, 2>{node_tags[2 * i], node_tags[2 * i + 1]});
  }
  return lines;
}

std::unordered_map<Tag, std::array<Tag, 3>> GmshInterface::GetFaces(
    const int tag) {
  std::vector<Tag> face_tags;
  std::vector<Tag> node_tags;
  gmsh::model::mesh::getElementsByType(
      static_cast<int>(ElementType::kThreeNodeTriangle), face_tags, node_tags,
      tag);

  // Insert the face and node tags into a map.
  std::unordered_map<Tag, std::array<Tag, 3>> faces;
  for (uint32_t i = 0; i < face_tags.size(); ++i) {
    faces.try_emplace(face_tags[i], std::array<gmsh::Tag, 3>{
                                        node_tags[3 * i], node_tags[3 * i + 1],
                                        node_tags[3 * i + 2]});
  }
  return faces;
}

std::unordered_map<Tag, std::array<Tag, 4>> GmshInterface::GetTetrahedra(
    const int tag) {
  std::vector<Tag> volume_tags;
  std::vector<Tag> node_tags;
  gmsh::model::mesh::getElementsByType(
      static_cast<int>(ElementType::kFourNodeTetrahedron), volume_tags,
      node_tags, tag);

  // Insert the volume and node tags into a map.
  std::unordered_map<Tag, std::array<Tag, 4>> volumes;
  for (uint32_t i = 0; i < volume_tags.size(); ++i) {
    volumes.try_emplace(
        volume_tags[i],
        std::array<gmsh::Tag, 4>{node_tags[4 * i], node_tags[4 * i + 1],
                                 node_tags[4 * i + 2], node_tags[4 * i + 3]});
  }
  return volumes;
}

std::vector<DimTag> GmshInterface::GetPhysicalGroups(const int dim) {
  std::vector<DimTag> dim_tags;
  gmsh::model::getPhysicalGroups(dim_tags, dim);
  return dim_tags;
}

std::vector<int> GmshInterface::GetEntitiesForPhysicalGroups(const int dim,
                                                             const int tag) {
  std::vector<int> entity_tags;
  gmsh::model::getEntitiesForPhysicalGroup(dim, tag, entity_tags);
  return entity_tags;
}

std::vector<int> GmshInterface::GetPhysicalGroupsForEntity(const int dim,
                                                           const int tag) {
  std::vector<int> physical_group_tags;
  gmsh::model::getPhysicalGroupsForEntity(dim, tag, physical_group_tags);
  return physical_group_tags;
}

}  // namespace gmsh

File: mesh/py/gmsh_interface.py
"""The Gmsh interface class provides an interface to gmsh utilities."""

import threading
from abc import ABC
from enum import Enum, IntEnum, auto

import gmsh
import numpy as np


class GmshElementType(IntEnum):
    """Mesh element type enumeration."""
    TWO_NODE_LINE = 1
    THREE_NODE_TRIANGLE = 2
    FOUR_NODE_QUADRANGLE = 3
    FOUR_NODE_TETRAHEDRON = 4
    EIGHT_NODE_HEXAHEDRON = 5
    SIX_NODE_PRISM = 6
    FIVE_NODE_PYRAMID = 7
    ONE_NODE_POINT = 15


class GmshNodeType(Enum):
    """Node type enumeration."""
    ALL = auto()
    INTERNAL = auto()
    BOUNDARY = auto()


class GmshInterface(ABC):
    """Gmsh interface."""

    _lifetime_lock = threading.Lock()
    _reference_count = 0

    def __init__(self) -> None:
        with GmshInterface._lifetime_lock:
            if GmshInterface._reference_count == 0:
                gmsh.initialize()
            GmshInterface._reference_count += 1

    def __del__(self) -> None:
        with GmshInterface._lifetime_lock:
            if GmshInterface._reference_count == 0:
                return
            GmshInterface._reference_count -= 1
            if GmshInterface._reference_count == 0:
                gmsh.finalize()

    @staticmethod
    def write_mesh_file(mesh_file: str) -> None:
        """Writes the mesh to the output file.

        Args:
            mesh_file: Mesh file.
        """
        gmsh.write(mesh_file)

    @staticmethod
    def launch() -> None:
        """Launches the GUI."""
        gmsh.fltk.run()

    @staticmethod
    def get_nodes(tag: int = -1,
                  dim: int = -1,
                  node_type: GmshNodeType = GmshNodeType.ALL) -> np.ndarray:
        """Returns the list of nodes for the tag and dimension.

        Args:
            tag: Tag of the entity.
            dim: Dimension of the nodes.
            node_type: The type of nodes to return.

        Returns:
            The list of node tags. The node tags may be duplicated across multiple
            dimensions.

        Raises:
            ValueError: If the node type is invalid.
        """

        node_tags, _, _ = gmsh.model.mesh.getNodes(tag=tag,
                                                   dim=dim,
                                                   includeBoundary=True)
        internal_node_tags, _, _ = gmsh.model.mesh.getNodes(
            tag=tag, dim=dim, includeBoundary=False)
        if node_type == GmshNodeType.ALL:
            return node_tags
        if node_type == GmshNodeType.INTERNAL:
            return internal_node_tags
        if node_type == GmshNodeType.BOUNDARY:
            return np.setdiff1d(node_tags, internal_node_tags)
        raise ValueError("Invalid node type.")

    @staticmethod
    def get_node_coordinates(
            tag: int = -1,
            dim: int = -1,
            coordinates_dim: int = 3) -> tuple[np.ndarray, np.ndarray]:
        """Returns the coordinates of the nodes.

        Args:
            tag: Tag of the entity.
            dim: Dimension of the nodes.
            coordinates_dim: Dimension of the coordinates.

        Returns:
            A 2-tuple consisting of the list of node tags and the list of
            corresponding coordinates.
        """
        node_tags, node_coordinates, _ = gmsh.model.mesh.getNodes(
            tag=tag, dim=dim, includeBoundary=True)
        return node_tags, np.reshape(node_coordinates,
                                     (-1, 3))[:, :coordinates_dim]

    @staticmethod
    def get_boundary(dim: int, tag: int) -> list[tuple[int, int]]:
        """Returns the boundary entities of the given entity.

        Args:
            dim: Dimension of the entity.
            tag: Tag of the entity.

        Returns:
            A list of 2-tuples consisting of the dimension and tag of each
            bounding entity.
        """
        return gmsh.model.getBoundary(dimTags=[(dim, tag)], oriented=False)

    @staticmethod
    def get_adjacencies(dim: int, tag: int) -> tuple[np.ndarray, np.ndarray]:
        """Returns the upward and downward adjacencies of the given entity.

        Args:
            dim: Dimension of the entity.
            tag: Tag of the entity.

        Returns:
            A 2-tuple consisting of the upward adjacent entities and the
            downward adjacent entities.
        """
        return gmsh.model.getAdjacencies(dim=dim, tag=tag)

    @staticmethod
    def get_lines(tag: int = -1) -> tuple[np.ndarray, np.ndarray]:
        """Returns the list of lines in the mesh.

        Args:
            tag: Tag of the entity.

        Returns:
            A 2-tuple consisting of the list of tags corresponding to the lines
            and the list of nodes adjacent to each line.
        """
        element_tags, node_tags = gmsh.model.mesh.getElementsByType(
            GmshElementType.TWO_NODE_LINE, tag=tag)
        return element_tags, np.reshape(node_tags, (-1, 2))

    @staticmethod
    def get_faces(tag: int = -1) -> tuple[np.ndarray, np.ndarray]:
        """Returns the list of triangular faces in the mesh.

        Args:
            tag: Tag of the entity.

        Returns:
            A 2-tuple consisting of the list of tags corresponding to the faces
            and the list of nodes belonging to each face.
        """
        element_tags, node_tags = gmsh.model.mesh.getElementsByType(
            GmshElementType.THREE_NODE_TRIANGLE, tag=tag)
        return element_tags, np.reshape(node_tags, (-1, 3))

    @staticmethod
    def get_tetrahedra(tag: int = -1) -> tuple[np.ndarray, np.ndarray]:
        """Returns the list of tetrahedral volumes in the mesh.

        Args:
            tag: Tag of the entity.

        Returns:
            A 2-tuple consisting of the list of tags corresponding to the
            tetrahedra and the list of nodes belonging to each volume.
        """
        element_tags, node_tags = gmsh.model.mesh.getElementsByType(
            GmshElementType.FOUR_NODE_TETRAHEDRON, tag=tag)
        return element_tags, np.reshape(node_tags, (-1, 4))

    @staticmethod
    def get_physical_groups(dim: int = -1) -> list[tuple[int, int]]:
        """Returns the list of physical groups in the mesh.

        Args:
            dim: Dimension of the physical groups.

        Returns:
            A 2-tuple consisting of the dimensions and the tags of the physical
            groups.
        """
        return gmsh.model.getPhysicalGroups(dim=dim)

    @staticmethod
    def get_entities_for_physical_group(dim: int, tag: int) -> list[int]:
        """Returns the list of entities making up the physical group.

        Args:
            dim: Dimension of the physical group.
            tag: Tag of the physical group.

        Returns:
            A list of tags corresponding to the entities making up the physical
            group.
        """
        return gmsh.model.getEntitiesForPhysicalGroup(dim=dim, tag=tag)

    @staticmethod
    def get_physical_groups_for_entity(dim: int, tag: int) -> list[int]:
        """Returns the list of physical groups to which the entity belongs.

        Args:
            dim: Dimension of the entity.
            tag: Tag of the entity.

        Returns:
            A list of tags corresponding to the physical groups.
        """
        return gmsh.model.getPhysicalGroupsForEntity(dim=dim, tag=tag)
